### PR TITLE
Fluence client

### PR DESCRIPTION
**Pull request by DieMyst**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -88,9 +88,12 @@ lazy val `cryptoJS` = `crypto`.js
     skip in packageJSDependencies := false
   )
 
+lazy val `client` = project.in(file("client"))
+  .dependsOn(`transport-grpc`, `kademlia-grpc`, `dataset-grpc`, `transport-core`, `kademlia-monix`, `dataset-protocol`)
+
 lazy val `dataset-node` = project.in(file("dataset/node"))
-  .dependsOn(`storage`, `kademlia-core`, `b-tree-server`, `kademlia-testkit` % Test, `dataset-client`, `b-tree-client`,
-              `dataset-client` % "compile->test")
+  .dependsOn(`storage`, `kademlia-monix`, `b-tree-server`, `kademlia-testkit` % Test, `dataset-client`, `b-tree-client`,
```
Comment (alari) [inline build.sbt]: Why does it depend on `kademlia-monix`? If it uses some monix-related utilities, why kademlia implementation comes along with them?

File: client/src/main/scala/fluence/client/AuthorizedClient.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import fluence.dataset.protocol.storage.DatasetStorageRpc
+import fluence.kad.protocol.Key
+import monix.eval.Task
+import scodec.bits.ByteVector
+
+import scala.util.Try
+
+case class AuthorizedClient(kp: KeyPair) {
+
+  def addDefaultDataset(storageRpc: DatasetStorageRpc[Task], clientState: Option[ClientState]) = {
```
Comment (alari) [inline client/src/main/scala/fluence/client/AuthorizedClient.scala]: What is it? What is "default" dataset? Why should it be added?

File: client/src/main/scala/fluence/client/AuthorizedClient.scala [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import fluence.crypto.keypair.KeyPair
+import fluence.dataset.client.ClientDatasetStorage
+import fluence.dataset.protocol.storage.DatasetStorageRpc
+import fluence.kad.protocol.Key
+import monix.eval.Task
+import scodec.bits.ByteVector
+
+import scala.util.Try
+
+case class AuthorizedClient(kp: KeyPair) {
```
Comment (alari) [inline client/src/main/scala/fluence/client/AuthorizedClient.scala]: What does this class do?

File: client/src/main/scala/fluence/client/FluenceClient.scala [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            k ← n.toKey(pk)
+            bc ← contracts.find(k)
+          } yield Left((n, l.updated(k, bc)))
+          t.onErrorHandleWith (_ ⇒ Task.pure(Right(l)))
+      }
+    }
+    findRec(Nonce(0), Map.empty)
+  }*/
+
+  def loadDatasetFromCache(pk: KeyPair.Public, dataStorage: Task[ClientDatasetStorage[String, String]]): Task[Option[ClientDatasetStorage[String, String]]] = {
```
Comment (alari) [inline client/src/main/scala/fluence/client/FluenceClient.scala]: Docs should be written. Why is it public? What is `dataStorage`? Where should it come from?

File: client/src/main/scala/fluence/client/FluenceClient.scala [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    datasetCache.getOrAddF(pk, dataStorage.map(Option.apply))
+  }
+
+  /**
+    * try to find dataset in cache,
+    * if it is absent, try to find contract in net and restore dataset from contract
+    * if it is no contract, create new contract and dataset
+    *
+    */
+  def restoreDataset(ac: AuthorizedClient): Task[ClientDatasetStorage[String, String]] = {
```
Comment (alari) [inline client/src/main/scala/fluence/client/FluenceClient.scala]: Should it be public? It looks like only `AuthorizedClient` should care about this method?

File: client/src/main/scala/fluence/client/FluenceClient.scala [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    val signer = signAlgo.signer(ac.kp)
+    for {
+      key ← Key.fromKeyPair(ac.kp)
+      bcOp ← contracts.find(key).attempt.map(_.toOption)
+      dataStorage ← bcOp match {
+        //create datastorage with old merkle root
+        case Some(bc) ⇒
+          for {
+            //TODO avoid _.head
+            ns ← kademlia.lookupIterative(bc.participants.head._1, 3)
```
Comment (alari) [inline client/src/main/scala/fluence/client/FluenceClient.scala]: You should use `findNode`

File: client/src/main/scala/fluence/client/FluenceClient.scala [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            offer ← BasicContract.offer(key, participantsRequired = 1, signer = signer)
+            accepted ← contracts.allocate(offer, dc ⇒ dc.sealParticipants(signer))
+            ns ← kademlia.findNode(accepted.participants.head._1, 10)
+            ds = ac.addDefaultDataset(storageRpc(ns.head.contact), None)
+          } yield ds
+      }
+    } yield dataStorage
+  }
+
+  def getOrCreateDataset(ac: AuthorizedClient): Task[ClientDatasetStorage[String, String]] = {
```
Comment (alari) [inline client/src/main/scala/fluence/client/FluenceClient.scala]: - Docs?
- Should it be public? Consider that this class is to be used by end users, so we should take care about the API, what is public and what is not, and try to keep API concise

File: client/src/main/scala/fluence/client/FluenceClient.scala [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  def generatePair(): Task[AuthorizedClient] = {
+    for {
+      kp ← signAlgo.generateKeyPair[Task]().value.flatMap(ME.fromEither)
+      ac ← loadOrCreateAuthorizedClient(kp)
+    } yield ac
+  }
+
+  def joinSeeds(contacts: List[Contact]): Task[Unit] = {
```
Comment (alari) [inline client/src/main/scala/fluence/client/FluenceClient.scala]: If `Kademlia` comes as DI into constructor, I suppose that joining seeds should be done somewhere above in the control flow? As several `FluenceClient`'s could be instantiated with the same `Kademlia` instance, while `join` should be launched only once?

File: client/src/main/scala/fluence/client/FluenceClient.scala [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    kademlia.join(contacts, 1)
+  }
+
+}
+
+object FluenceClient {
+  def apply(
```
Comment (alari) [inline client/src/main/scala/fluence/client/FluenceClient.scala]: Please write docs. Should it be called for each dataset? Should be different Kademlia provided for each instance, or could it be shared? Is it the main entry point to using Fluence, or should there be another layer above it?

File: crypto/src/main/scala/fluence/crypto/keypair/KeyPair.scala [hunk 10]
```diff
@@ -19,6 +19,7 @@ package fluence.crypto.keypair
 
 import scodec.bits.ByteVector
 
+//TODO store secret key safer
```
Comment (alari) [inline crypto/src/main/scala/fluence/crypto/keypair/KeyPair.scala]: What do you mean?

File: kademlia/monix/src/main/scala/fluence/kad/MVarMapCache.scala [hunk 11]
```diff
@@ -0,0 +1,82 @@
+package fluence.kad
+
+import cats.data.StateT
+import monix.eval.{ MVar, Task }
+
+import scala.collection.concurrent.TrieMap
+
+//todo move this to some utility module, replace all similar caches with MVarCache
+//todo we always return non-default parameters, prove it on types
+class MVarMapCache[K, V](default: V) {
```
Comment (alari) [inline kademlia/monix/src/main/scala/fluence/kad/MVarMapCache.scala]: Docs should be written both for class and its methods, as it's unclear why it exists and what problems it solves

File: kademlia/monix/src/test/scala/fluence/kad/MVarCacheTest.scala [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+      future.onComplete {
+        case Success(_)         ⇒ ()
+        case Failure(exception) ⇒ println(Console.RED + s"TASK ERROR: $exception")
+      }(s)
+      future.futureValue
+    }
+  }
+
+  "some" should {
+    "work" in {
```
Comment (alari) [inline kademlia/monix/src/test/scala/fluence/kad/MVarCacheTest.scala]: Probably we need some concurrent-intensive tests for it?

File: client/src/main/scala/fluence/client/FluenceClient.scala [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+            ns ← kademlia.lookupIterative(bc.participants.head._1, 3)
+            ds = ac.addDefaultDataset(storageRpc(ns.head.contact), Some(ClientState(bc.executionState.merkleRoot.toArray)))
+          } yield ds
+        //new datastorage
+        case None ⇒
+          for {
+            //TODO avoid _.head
+            offer ← BasicContract.offer(key, participantsRequired = 1, signer = signer)
+            accepted ← contracts.allocate(offer, dc ⇒ dc.sealParticipants(signer))
+            ns ← kademlia.findNode(accepted.participants.head._1, 10)
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/FluenceClient.scala]: why 10? not 1 or 30?

File: dataset/client/src/main/scala/fluence/dataset/client/ClientDatasetStorage.scala [hunk 14]
```diff
@@ -26,6 +26,7 @@ import monix.eval.Task
 
 import scala.language.higherKinds
 
+//todo datasetId - kademlia key
```
Comment (RicoGit) [inline dataset/client/src/main/scala/fluence/dataset/client/ClientDatasetStorage.scala]: Yes, it should be more clear!

File: kademlia/monix/src/test/scala/fluence/kad/MVarCacheTest.scala [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+import monix.eval.Task
+import monix.execution.Scheduler.Implicits.global
+import monix.execution.{ CancelableFuture, Scheduler }
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.{ Matchers, WordSpec }
+
+import scala.util.{ Failure, Success }
+
+class MVarCacheTest extends WordSpec with Matchers with ScalaFutures {
```
Comment (RicoGit) [inline kademlia/monix/src/test/scala/fluence/kad/MVarCacheTest.scala]: Test -> Spec

File: kademlia/monix/src/test/scala/fluence/kad/MVarCacheTest.scala [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+  private implicit class WaitTask[T](task: Task[T]) {
+    def taskValue()(implicit s: Scheduler): T = {
+      val future: CancelableFuture[T] = task.runAsync(s)
+      future.onComplete {
+        case Success(_)         ⇒ ()
+        case Failure(exception) ⇒ println(Console.RED + s"TASK ERROR: $exception")
+      }(s)
+      future.futureValue
+    }
+  }
```
Comment (RicoGit) [inline kademlia/monix/src/test/scala/fluence/kad/MVarCacheTest.scala]: Looks so familiar)

File: node/src/test/scala/fluence/node/ClientNodeIntegrationSpec.scala [hunk 17]
```diff
@@ -315,7 +354,7 @@ class ClientNodeIntegrationSpec extends WordSpec with Matchers with ScalaFutures
     val value1: Option[ClientState] = merkleRoot.map(ClientState)
     ClientDatasetStorage(
       datasetId,
-      TestCryptoHasher,
+      JdkCryptoHasher.Sha256,
```
Comment (RicoGit) [inline node/src/test/scala/fluence/node/ClientNodeIntegrationSpec.scala]: Why did you do that? I prefer use TestHasher in test. You should parametrize kademlia Client with hasher, to be able to change hasher.

File: node/src/test/scala/fluence/node/ClientNodeIntegrationSpec.scala [hunk 18]
```diff
@@ -282,22 +283,60 @@ class ClientNodeIntegrationSpec extends WordSpec with Matchers with ScalaFutures
...
+        ds.put("jey2", "esrk2").taskValue()
+        ds.put("jey3", "esrk3").taskValue()
+
+        //todo this should be equals, but something wrong
+        ds.get("jey1").taskValue()
+        ds.get("jey2").taskValue()
+        ds.get("jey3").taskValue()
+      }
+    }
+
```
Comment (RicoGit) [inline node/src/test/scala/fluence/node/ClientNodeIntegrationSpec.scala]: Let's remove this test, in my branch there is the same, but more complex test.

File: client/src/main/scala/fluence/client/FluenceClient.scala [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+  /**
+   * Create string dataset with noop-encryption
+   * private until multiple datasets per authorized client is supported
+   * @param contact node where we will store dataset
+   * @param clientState merkle root if it is not a new dataset
+   * @return dataset representation
+   */
+  private def addNonEncryptedDataset(ac: AuthorizedClient, contact: Contact, clientState: Option[ClientState]): Task[ClientDatasetStorage[String, String]] = {
+    addDataset(ac, storageRpc(contact), NoOpCrypt.forString, NoOpCrypt.forString, clientState, JdkCryptoHasher.Sha256)
+  }
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/FluenceClient.scala]: JdkCryptoHasher.Sha256 should not be hardcoded

File: client/src/main/scala/fluence/client/FluenceClient.scala [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+
+class FluenceClient(
+    kademlia: Kademlia[Task, Contact],
+    contracts: Contracts[Task, BasicContract, Contact],
+    signAlgo: SignAlgo,
+    storageRpc: Contact ⇒ DatasetStorageRpc[Task]
+)(implicit ME: MonadError[Task, Throwable]) {
+
+  val authorizedCache = TrieMap.empty[KeyPair.Public, AuthorizedClient]
+  val datasetCache = TrieMap.empty[KeyPair.Public, Task[ClientDatasetStorage[String, String]]]
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/FluenceClient.scala]: Perhaps, authorizedCache and datasetCache should be private.

**Pull request merged**